roadmap/implementers-guide/src/types/network.md

```diff
@@ -154,7 +154,10 @@ enum NetworkBridgeEvent<M> {
 	///
 	/// Note, that the distribution subsystems need to handle the last
 	/// view update of the newly added gossip peers manually.
-	NewGossipTopology(HashSet<PeerId>),
+	NewGossipTopology {
+		our_neighbors_x: HashSet<PeerId>,
+		our_neighbors_y: HashSet<PeerId>,
+	}
```

Comment:
Since we need to answer "given ValidatorIndex is it our row or column neighbor and if so what are PeerIds of the column/row neighbors accordingly?", we probably should store some more information about indices. It can be something like
```rust
struct GridNeighbors {
    our_neighbors_x: bimap::hash::BiHashMap<PeerId, ValidatorIndex>,
    our_neighbors_y: bimap::hash::BiHashMap<PeerId, ValidatorIndex>,
    <snip>
}
```
This assume one `PeerId` per validator (cf. https://github.com/paritytech/substrate/pull/10259#discussion_r748851902). Should we account for a validator changing its `PeerId`? That would complicate things.

Reply:
> Since we need to answer "given ValidatorIndex is it our row or column neighbor and if so what are PeerIds of the column/row neighbors accordingly?"

Yes - updated. But we actually need to answer "give ValidatorIndex and session" which necessitated making gossip-support session aware. Also, the previous behavior of gossip-support didn't really make sense with validators rotating their keys.